kroxylicious-schema-validation/src/main/java/io/kroxylicious/proxy/filter/schema/ProduceValidationFilter.java

```diff
@@ -0,0 +1,122 @@
+/*
+ * Copyright Kroxylicious Authors.
+ *
+ * Licensed under the Apache Software License version 2.0, available at http://www.apache.org/licenses/LICENSE-2.0
+ */
+package io.kroxylicious.proxy.filter.schema;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import org.apache.kafka.common.message.ProduceRequestData;
+import org.apache.kafka.common.message.ProduceResponseData;
+import org.apache.kafka.common.message.RequestHeaderData;
+import org.apache.kafka.common.message.ResponseHeaderData;
+import org.apache.kafka.common.protocol.Errors;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import io.kroxylicious.proxy.filter.KrpcFilterContext;
+import io.kroxylicious.proxy.filter.ProduceRequestFilter;
+import io.kroxylicious.proxy.filter.ProduceResponseFilter;
+import io.kroxylicious.proxy.filter.schema.validation.ProduceRequestValidator;
+import io.kroxylicious.proxy.filter.schema.validation.RecordValidationFailure;
+import io.kroxylicious.proxy.filter.schema.validation.TopicValidationResult;
+import io.kroxylicious.proxy.filter.schema.validation.ValidationResult;
+
+import static java.util.stream.Collectors.joining;
+
+public class ProduceValidationFilter
+        implements ProduceRequestFilter, ProduceResponseFilter {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(ProduceValidationFilter.class);
+    private final ProduceRequestValidator validator;
+    private final Map<Integer, ValidationResult> correlatedResults = new HashMap<>();
+
+    public ProduceValidationFilter(ProduceRequestValidator validator) {
+        this.validator = validator;
+    }
+
+    @Override
+    public void onProduceRequest(RequestHeaderData header, ProduceRequestData request, KrpcFilterContext context) {
+        ValidationResult result = validator.validateRequest(request);
+        if (result.allTopicPartitionsInvalid()) {
+            LOGGER.warn("all topic-partitions for request contained invalid data: {}", result);
+            ProduceResponseData response = invalidateAllTopicPartitionProduceRequests(result);
+            context.forwardResponse(response);
+        }
+        else if (result.anyTopicPartitionInvalid()) {
```

Comment:
The normal case, and the one which we want to optimize for, will be that batches contain only valid records. So perhaps we could refactor this `if/else if/else` to check only for the presence of some invalid records on the happy path, and only check for the all invalid records case once that check has failed. I.e. in pseudocode:

```
if (someInvalid(batch)) {
  if (allInvalid(batch)) {
    context.forwardResponse()
  } else {
    context.forwardRequest()
  }
} else {
    context.forwardRequest()
}
```